crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -977,11 +996,34 @@ impl<'db> TypeInferenceBuilder<'db> {
         // the runtime error will occur immediately (rather than when the symbol is *used*,
         // as would be the case for a symbol with type `Unbound`), so it's appropriate to
         // think of the type of the imported symbol as `Unknown` rather than `Unbound`
-        let ty = module_ty
+        let member_ty = module_ty
+            .unwrap_or(Type::Unbound)
             .member(self.db, &Name::new(&name.id))
```

Comment:
I still find it odd that we have to do the "unbound" check everywhere where we call `.member`. Do we not always want to emit a diagnostic when failing to resolve a member?

Reply:
We do, but it will need to be different error codes. It would be very counterintuitive for users if `import foo` triggers an `unresolved-import` error code and `from foo import bar` triggers an `unresolved-member` error code. Users will expect all import-related failures to be the same error code and all attribute-access-related failures to be a different error code.

Reply:
That makes sense. We might just want to parameterize the `member` method. But maybe that's not worth it? I don't know. Have you looked at how mypy/pyright do this?

Reply:
> We might just want to parameterize the `member` method.

Not sure exactly what you mean; could you clarify?

> Have you looked at how mypy/pyright do this?

Mypy has a very different architecture to the one we're building. It resolves all imports eagerly in a first pass before it does any other type checking; `module-not-found` errors are emitted during this first pass here: <https://github.com/python/mypy/blob/fe15ee69b9225f808f8ed735671b73c31ae1bed8/mypy/build.py#L2600-L2700>.

I'm not as familiar with pyright's codebase but I can dig in now

Reply:
Hmm... it seems both mypy and pyright in fact do what you'd prefer here -- they both use "attribute-access" error codes for `from x import y` imports:
- pyright: <https://pyright-play.net/?pyrightVersion=1.1.368&code=GYJw9gtgBAxmA28CmMAuBLMA7AzgOgEMAjGKdCABzBFSmDDCA>
- mypy: <https://mypy-play.net/?mypy=latest&python=3.12&gist=3110174b575f8d8eba319462c45a8ddf>

Pyright's error code is `reportAttributeAccessIssue`; mypy's is `attr-defined`, which both seem equally bad. In terms of error messages, pyright definitely wins, though: pyright has:

```
"foo" is unknown import symbol
```

whereas mypy has

```
Module "collections.abc" has no attribute "foo" 
```

Given this precedent, I'm okay with emitting the diagnostic from `.member()`, if you'd prefer. Though I hope we still aspire to be more user-friendly in our error messages than either mypy or pyright.

Reply:
I'm not suggesting that we should use a simple error message. I'm only raising the question if we should change `member` so that it always requires handling the error where the member might be unbound by either returning a `Result` or having a closure or similar to create a diagnostic if the member is unbound.

Reply:
Oh, I agree that that makes a lot of sense! I can make that change

Reply:
Hmmm... if the inferred type of a member access is a union, e.g. `int | Unbound`, would you expect that to be represented as an `Ok()` or an `Err()` variant (if we return an `Result` from `.member()`?

Reply:
> having a closure or similar to create a diagnostic if the member is unbound.

This might be doable, but I think it would still be quite a large refactor and design change. I think it's out of the scope of this PR.

Reply:
I'm not convinced there is any problem with reporting an error on `from foo import bar` (where the module `foo` does exist, but has no top-level symbol `bar`) as an attribute-access error code. If I had to pick between the pyright or mypy error messages above, I would prefer the mypy one (the pyright one gives less useful context). The fact that the attribute access occurred as part of a from-import will be clear from the error location; it doesn't have to be part of the error message or code. I wouldn't want to make design decisions based on the hypothesis that this is confusing to users without clear evidence that it is (e.g. a history of users filing issues on mypy about that error code or message); personally I find it intuitive.